KAFKA-10173: Fix suppress changelog binary schema compatibility

vvcephei opened the pull request:
We inadvertently changed the binary schema of the suppress buffer changelog
in 2.4.0 without bumping the schema version number. As a result, it is impossible
to upgrade from 2.3.x to 2.4+ if you are using suppression.

* Refactor the schema compatibility test to use serialized data from older versions
as a more foolproof compatibility test.
* Refactor the upgrade system test to use the smoke test application so that we
actually exercise a significant portion of the Streams API during upgrade testing
* Add more recent versions to the upgrade system test matrix
* Fix the compatibility bug by bumping the schema version to 3

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

Commit:
KAFKA-10173: Directly use Arrays.equals for version comparison

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -286,6 +289,15 @@ private void logTombstone(final Bytes key) {
 
     private void restoreBatch(final Collection<ConsumerRecord<byte[], byte[]>> batch) {
         for (final ConsumerRecord<byte[], byte[]> record : batch) {
+            if (record.partition() != partition) {
+                throw new IllegalStateException(
+                    String.format(
+                        "record partition [%d] is being restored by the wrong suppress partition [%d]",
+                        record.partition(),
+                        partition
+                    )
+                );
+            }
```
On the side, I realized we can consolidate this check and perform it first, rather than after we're already written bad data into the buffer.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -339,7 +341,7 @@ private void restoreBatch(final Collection<ConsumerRecord<byte[], byte[]>> batch
                         )
                     );
-                } else if (V_1_CHANGELOG_HEADERS.lastHeader("v").equals(record.headers().lastHeader("v"))) {
+                } else if (Arrays.equals(record.headers().lastHeader("v").value(), V_1_CHANGELOG_HEADER_VALUE)) {
```
This is the fix (although it was probably fine before). The implementation of Header.equals is not specified by any contract, so it's safer to perform a direct comparison on the header values. Just as before, I'm comparing byte arrays to avoid deserializing the value.

chia7712 commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -58,10 +59,12 @@
     private static final BytesSerializer KEY_SERIALIZER = new BytesSerializer();
     private static final ByteArraySerializer VALUE_SERIALIZER = new ByteArraySerializer();
+    private static final byte[] V_1_CHANGELOG_HEADER_VALUE = {(byte) 1};
     private static final RecordHeaders V_1_CHANGELOG_HEADERS =
```
my IDEA says this variable is never used.

guozhangwang replied:
I saw it is used in line 342.

chia7712 replied:
my bad. The unused variable is V_1_CHANGELOG_HEADERS rather than V_1_CHANGELOG_HEADER_VALUE

vvcephei replied:
Ah, right. My mistake. Thanks for pointing it out.

chia7712 commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -299,16 +311,6 @@ private void restoreBatch(final Collection<ConsumerRecord<byte[], byte[]>> batch
-                    );
-                }
             } else {
                 if (record.headers().lastHeader("v") == null) {
```
nit:
We seek the last header many times. Could we reuse the return value?

vvcephei replied:
Sure!

guozhangwang reviewed (commented):
The PR lgtm, please feel free to merge after addressed @chia7712 's comments.

Commit:
converting upgrade test to smoke test

Commit:
wip debugging suppress buffer

Commit:
asdf

Commit:
fix

Commit:
upgrade test passed (2.3.1 -> trunk)

Commit:
wip

Commit:
cleanup

Commit:
cleanup

Commit:
fix test

Commit:
direct encoding of restore test vx

Commit:
direct encoding of restore test v1

Commit:
add v2 and v3

Commit:
cleanup

vvcephei commented:
I'm still cleaning up this PR. I'll call for reviews when it's ready.

Commit:
adding other old versions

Commit:
fallback to make 2.4 work

Commit:
remove 2.1. cf KAFKA-10203

Commit:
reduce cyclomatic complexity

Commit:
Remove ParallelGC jvm param

I was getting this exception, and somehow, the parallel GC parameter was the culprit

    java.lang.OutOfMemoryError: Java heap space
        at org.apache.kafka.streams.kstream.internals.FullChangeSerde.decomposeLegacyFormattedArrayIntoChangeArrays(FullChangeSerde.java:82)
        at org.apache.kafka.streams.state.internals.TimeOrderedKeyValueBufferChangelogDeserializationHelper.deserializeV2(TimeOrderedKeyValueBufferChangelogDeserializationHelper.java:90)
        at org.apache.kafka.streams.state.internals.TimeOrderedKeyValueBufferChangelogDeserializationHelper.duckTypeV2(TimeOrderedKeyValueBufferChangelogDeserializationHelper.java:61)
        at org.apache.kafka.streams.state.internals.InMemoryTimeOrderedKeyValueBuffer.restoreBatch(InMemoryTimeOrderedKeyValueBuffer.java:369)
        at org.apache.kafka.streams.state.internals.InMemoryTimeOrderedKeyValueBuffer$$Lambda$284/0x00000001002cb440.restoreBatch(Unknown Source)
        at org.apache.kafka.streams.state.internals.TimeOrderedKeyValueBufferTest.shouldRestoreV3FormatWithV2Header(TimeOrderedKeyValueBufferTest.java:742)

Commit:
cleanup

vvcephei commented on build.gradle:
```diff
@@ -97,7 +97,7 @@ ext {
 
   defaultMaxHeapSize = "2g"
-  defaultJvmArgs = ["-Xss4m", "-XX:+UseParallelGC"]
+  defaultJvmArgs = ["-Xss4m"]
```
@ijuma , you'll probably want to know about this.

I have no idea why, but one of the new tests in this PR was failing with:
```
    java.lang.OutOfMemoryError: Java heap space
        at org.apache.kafka.streams.kstream.internals.FullChangeSerde.decomposeLegacyFormattedArrayIntoChangeArrays(FullChangeSerde.java:82)
        at org.apache.kafka.streams.state.internals.TimeOrderedKeyValueBufferChangelogDeserializationHelper.deserializeV2(TimeOrderedKeyValueBufferChangelogDeserializationHelper.java:90)
        at org.apache.kafka.streams.state.internals.TimeOrderedKeyValueBufferChangelogDeserializationHelper.duckTypeV2(TimeOrderedKeyValueBufferChangelogDeserializationHelper.java:61)
        at org.apache.kafka.streams.state.internals.InMemoryTimeOrderedKeyValueBuffer.restoreBatch(InMemoryTimeOrderedKeyValueBuffer.java:369)
        at org.apache.kafka.streams.state.internals.InMemoryTimeOrderedKeyValueBuffer$$Lambda$284/0x00000001002cb440.restoreBatch(Unknown Source)
        at org.apache.kafka.streams.state.internals.TimeOrderedKeyValueBufferTest.shouldRestoreV3FormatWithV2Header(TimeOrderedKeyValueBufferTest.java:742)
```

I captured a flight recording and a heap dump on exit, but everything looked fine, and the heap was only a few megs at the time of the crash. I noticed first that if I just overrode all the jvm args, the test would pass, and through trial and error, I identified this one as the "cause".

I get an OOMe every time with `-XX:+UseParallelGC` and I've never gotten it without the flag. WDYT about dropping it?

vvcephei replied:
Aha! I figured it out. There actually was a bug in the test. While duck-typing, the code was trying to allocate an array of 1.8GB. It's funny that disabling this flag made this test pass on java 11 and 14. Maybe the flag partitions the heap on those versions or something, so the test didn't actually have the full 2GB available. Anyway, I'm about to push a fix and put the flag back.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/FullChangeSerde.java:
```diff
@@ -68,33 +68,6 @@ private FullChangeSerde(final Serde<T> inner) {
-     * We used to serialize a Change into a single byte[]. Now, we don't anymore, but we still keep this logic here
-     * so that we can produce the legacy format to test that we can still deserialize it.
-     */
-    public static byte[] mergeChangeArraysIntoSingleLegacyFormattedArray(final Change<byte[]> serialChange) {
```
Only used in the test now, so I moved it.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/BufferValue.java:
```diff
@@ -120,7 +120,7 @@ ByteBuffer serialize(final int endPadding) {
         if (oldValue == null) {
             buffer.putInt(NULL_VALUE_SENTINEL);
-        } else if (priorValue == oldValue) {
+        } else if (Arrays.equals(priorValue, oldValue)) {
```
This was correct before, since we check equality and enforce identity in the constructor, but `Arrays.equals` is extremely cheap when the arrays are identical, so explicitly doing an identity check instead of equality was a micro-optimization.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -54,14 +56,17 @@
-        new RecordHeaders(new Header[] {new RecordHeader("v", new byte[] {(byte) 1})});
-    private static final RecordHeaders V_2_CHANGELOG_HEADERS =
-        new RecordHeaders(new Header[] {new RecordHeader("v", new byte[] {(byte) 2})});
+    private static final byte[] V_1_CHANGELOG_HEADER_VALUE = {(byte) 1};
+    private static final byte[] V_2_CHANGELOG_HEADER_VALUE = {(byte) 2};
+    private static final byte[] V_3_CHANGELOG_HEADER_VALUE = {(byte) 3};
+    static final RecordHeaders CHANGELOG_HEADERS =
+        new RecordHeaders(new Header[] {new RecordHeader("v", V_3_CHANGELOG_HEADER_VALUE)});
```
We don't need to store the whole RecordHeaders for the old versions, just the actual version flag.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -361,26 +366,20 @@ private void restoreBatch(final Collection<ConsumerRecord<byte[], byte[]>> batch
-                    // in this case, the changelog value is a serialized BufferValue
+                } else if (Arrays.equals(versionHeader.value(), V_2_CHANGELOG_HEADER_VALUE)) {
+
+                    final DeserializationResult deserializationResult = duckTypeV2(record, key);
```
See the comment on this method for why we need to duck-type version 2. I pulled these deserializations into a helper class because all the extra branches pushed our cyclomatic complexity over the limit.

But I kept the first two branches here because they aren't pure functions. They perform a lookup in the buffer itself as part of converting the old format.

guozhangwang replied:
Could you clarify which comment are you referring to? I did not see any comments for the "restoreBatch" method..

vvcephei replied:
Sorry, the comments in `duckTypeV2`.

Basically, because we released three versions that would write data in the "v3" format, but with the "v2" flag, when we see the v2 flag, the data might be in v2 format or v3 format. The only way to tell is to just try to deserialize it in v2 format, and if we get an exception, then to try with v3 format.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -481,8 +480,7 @@ public void put(final long time,
         if (buffered == null) {
-            final V priorValue = value.oldValue;
-            serializedPriorValue = (priorValue == null) ? null : valueSerde.innerSerde().serializer().serialize(changelogTopic, priorValue);
+            serializedPriorValue = serialChange.oldValue;
```
If you look a few lines up, you'll see that we just serialized the "old value", so we don't need to serialize it again here.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/integration/SmokeTestDriverIntegrationTest.java:
```diff
@@ -104,10 +104,6 @@ public void shouldWorkWithRebalance() throws InterruptedException {
             clients.add(smokeTestClient);
             smokeTestClient.start(props);
 
-            while (!clients.get(clients.size() - 1).started()) {
-                Thread.sleep(100);
-            }
-
```
Don't need this anymore because `start` blocks until it's "started" now.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/FullChangeSerdeTest.java:
```diff
@@ -19,17 +19,46 @@
+     * We used to serialize a Change into a single byte[]. Now, we don't anymore, but we still keep this logic here
+     * so that we can produce the legacy format to test that we can still deserialize it.
+     */
+    private static byte[] mergeChangeArraysIntoSingleLegacyFormattedArray(final Change<byte[]> serialChange) {
```
Moved from FullChangeSerde because it is only used in this test.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedKeyValueBufferTest.java:
```diff
@@ -56,14 +55,13 @@
 
 @RunWith(Parameterized.class)
 public class TimeOrderedKeyValueBufferTest<B extends TimeOrderedKeyValueBuffer<String, String>> {
-    private static final RecordHeaders V_2_CHANGELOG_HEADERS =
-        new RecordHeaders(new Header[] {new RecordHeader("v", new byte[] {(byte) 2})});
```
imported the headers from the production code, so that it'll stay current.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedKeyValueBufferTest.java:
```diff
@@ -372,12 +370,14 @@ public void shouldRestoreOldFormat() {
         context.setRecordContext(new ProcessorRecordContext(0, 0, 0, "", null));
 
-        final FullChangeSerde<String> serializer = FullChangeSerde.wrap(Serdes.String());
+        // These serialized formats were captured by running version 2.1 code.
+        // They verify that an upgrade from 2.1 will work.
+        // Do not change them.
+        final String toDeleteBinaryValue = "0000000000000000FFFFFFFF00000006646F6F6D6564";
+        final String asdfBinaryValue = "0000000000000002FFFFFFFF0000000471776572";
+        final String zxcvBinaryValue1 = "00000000000000010000000870726576696F757300000005656F34696D";
+        final String zxcvBinaryValue2 = "000000000000000100000005656F34696D000000046E657874";
```
This was one of my major findings in KAFKA-10173. Because the test was serializing the "old versions" using code shared with the current logic, we could not detect when we accidentally changed the current serialization logic without bumping the version number.

By instead testing against fixed pre-serialized data, we should be a lot safer.

I took inspiration from the way that Karsten reported the observed serialized data in the bug report. Hex-encoding the binary data makes the tests more readable than a long array of byte literals.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -361,26 +366,20 @@ private void restoreBatch(final Collection<ConsumerRecord<byte[], byte[]>> batch
+                    final DeserializationResult deserializationResult = duckTypeV2(record, key);
+                    cleanPut(deserializationResult.time(), deserializationResult.key(), deserializationResult.bufferValue());
+
+                } else if (Arrays.equals(versionHeader.value(), V_3_CHANGELOG_HEADER_VALUE)) {
```
This is the version bump we should have done in 2.4.0. I'll backport this fix to the 2.4 branch.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestClient.java:
```diff
@@ -38,107 +37,128 @@
-    public SmokeTestClient(final String name) {
-        super();
-        this.name = name;
+    private static void addShutdownHook(final String name, final Runnable runnable) {
```
I inlined these utilities to make this class more "portable". I.e., so that we can copy-paste it into the upgrade-test modules without dragging in a bunch of extra dependencies.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestClient.java:
```diff
@@ -38,107 +37,128 @@
 
     public void start(final Properties streamsProperties) {
-        streams = createKafkaStreams(streamsProperties);
+        final Topology build = getTopology();
```
This was another bug I happened to notice while scrutinizing this system test. `createKafkaStreams` was registering a state listener and exception handler. But the next line here was overriding the exception handler, so the one registered in `createKafkaStreams` was getting ignored. I noticed it because I registered a state listener here, which also caused the one registered in `createKafkaStreams` to get ignored.

Inlining solves this problem, and since `createKafkaStreams` had only one usage, it was needless anyway.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestClient.java:
```diff
@@ -38,107 +37,128 @@
+            System.out.println(name + ": SMOKE-TEST-CLIENT-EXCEPTION: " + e);
+            e.printStackTrace(System.out);
+        }
+        System.out.println(name + ": SMOKE-TEST-CLIENT-STARTED");
```
A new message we can look for to wait until the instance has completed joining the group.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestClient.java:
```diff
@@ -38,107 +37,128 @@
-        if (!uncaughtException) {
+        final boolean wasClosed = streams.close(Duration.ofMinutes(1));
+
+        if (wasClosed && !uncaughtException) {
```
Found a missed condition, if the close timed out, there wouldn't be an exception, just a false return value.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestClient.java:
```diff
@@ -38,107 +37,128 @@
         final Properties fullProps = new Properties(props);
         fullProps.put(StreamsConfig.APPLICATION_ID_CONFIG, "SmokeTest");
         fullProps.put(StreamsConfig.CLIENT_ID_CONFIG, "SmokeTest-" + name);
-        fullProps.put(StreamsConfig.NUM_STREAM_THREADS_CONFIG, 3);
-        fullProps.put(StreamsConfig.NUM_STANDBY_REPLICAS_CONFIG, 2);
-        fullProps.put(StreamsConfig.BUFFERED_RECORDS_PER_PARTITION_CONFIG, 100);
-        fullProps.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 1000);
-        fullProps.put(StreamsConfig.REPLICATION_FACTOR_CONFIG, 3);
-        fullProps.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
-        fullProps.put(ProducerConfig.ACKS_CONFIG, "all");
```
I moved all these to the system test `propFile()` definition.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestClient.java:
```diff
@@ -38,107 +37,128 @@
         return fullProps;
     }
 
-    private KafkaStreams createKafkaStreams(final Properties props) {
```
inlined above.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestUtil.java:
```diff
@@ -75,7 +75,7 @@ public void process(final Object key, final Object value) {
                     @Override
                     public void close() {
-                        System.out.printf("Close processor for task %s", context().taskId());
+                        System.out.printf("Close processor for task %s%n", context().taskId());
```
Oops! I noticed the lack of a newline in the output. It didn't matter for the tests because the greps aren't bounded by line.

vvcephei commented on streams/upgrade-system-tests-22/src/test/java/org/apache/kafka/streams/tests/SmokeTestClient.java:
```diff
@@ -0,0 +1,298 @@
+
+import static org.apache.kafka.streams.kstream.Suppressed.untilWindowCloses;
+
+public class SmokeTestClient extends SmokeTestUtil {
```
All of these are just copy-pasted from the main module.

vvcephei commented on tests/kafkatest/services/streams.py:
```diff
@@ -305,23 +305,62 @@ def start_node(self, node):
     """Base class for Streams Smoke Test services providing some common settings and functionality"""
 
-    def __init__(self, test_context, kafka, command, processing_guarantee = 'at_least_once', num_threads = 3):
+    def __init__(self, test_context, kafka, command, processing_guarantee = 'at_least_once', num_threads = 3, replication_factor = 3):
```
Adding a param so that we can get away with just one broker in the upgrade test.

vvcephei commented on tests/kafkatest/services/streams.py:
```diff
@@ -305,23 +305,62 @@ def start_node(self, node):
+        self.UPGRADE_FROM = None
+        self.REPLICATION_FACTOR = replication_factor
+
+    def set_version(self, kafka_streams_version):
+        self.KAFKA_STREAMS_VERSION = kafka_streams_version
+
+    def set_upgrade_from(self, upgrade_from):
+        self.UPGRADE_FROM = upgrade_from
```
Added for the upgrade test.

vvcephei commented on tests/kafkatest/services/streams.py:
```diff
@@ -305,23 +305,62 @@ def start_node(self, node):
                       streams_property.PROCESSING_GUARANTEE: self.PROCESSING_GUARANTEE,
-                      streams_property.NUM_THREADS: self.NUM_THREADS}
+                      streams_property.NUM_THREADS: self.NUM_THREADS,
+                      "replication.factor": self.REPLICATION_FACTOR,
+                      "num.standby.replicas": 2,
+                      "buffered.records.per.partition": 100,
+                      "commit.interval.ms": 1000,
+                      "auto.offset.reset": "earliest",
+                      "acks": "all"}
```
Moved from the java code so that all the configs can be defined together.

vvcephei commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -37,6 +37,9 @@
 # can be replaced with metadata_2_versions
 backward_compatible_metadata_2_versions = [str(LATEST_0_10_2), str(LATEST_0_11_0), str(LATEST_1_0), str(LATEST_1_1)]
 metadata_3_or_higher_versions = [str(LATEST_2_0), str(LATEST_2_1), str(LATEST_2_2), str(LATEST_2_3), str(LATEST_2_4), str(LATEST_2_5), str(DEV_VERSION)]
+smoke_test_versions = [str(LATEST_2_2), str(LATEST_2_3), str(LATEST_2_4), str(LATEST_2_5)]
```
See KAFKA-10203 for why I couldn't go past 2.2

vvcephei commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -189,8 +192,8 @@ def test_upgrade_downgrade_brokers(self, from_version, to_version):
 
-    @matrix(from_version=metadata_2_versions, to_version=metadata_2_versions)
-    def test_simple_upgrade_downgrade(self, from_version, to_version):
+    @matrix(from_version=smoke_test_versions, to_version=dev_version)
```
We were previously not testing 2.0+ _at all_. After rewriting this as a smoke test, it only applies to 2.2+. I also figured it makes more sense just to test upgrades to the current branch, rather than testing cross-upgrades between every pair of versions.

guozhangwang replied:
+1, I think this is a great find.

vvcephei commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -201,14 +204,29 @@ def test_simple_upgrade_downgrade(self, from_version, to_version):
         self.zk.start()
 
-        self.kafka = KafkaService(self.test_context, num_nodes=1, zk=self.zk, topics=self.topics)
+        self.kafka = KafkaService(self.test_context, num_nodes=1, zk=self.zk, topics={
```
A lot of these changes are part of adapting the test to the smoke test app.

vvcephei commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -349,56 +370,42 @@ def get_version_string(self, version):
     def start_all_nodes_with(self, version):
```
I refactored this method to start all the nodes concurrently, rather than one at a time. We still do a rolling upgrade, but there's no need to do a rolling startup.

vvcephei reviewed (commented):
Whew! Can you take another look @guozhangwang and @chia7712 ?

After finding the root cause, I was able to fixed several related problems. The diff size is unfortunate but it's almost all the result of copy/pasting the smoke test into the upgrade-system-tests modules.

guozhangwang commented on streams/upgrade-system-tests-25/src/test/java/org/apache/kafka/streams/tests/StreamsSmokeTest.java:
```diff
@@ -0,0 +1,99 @@
+import static org.apache.kafka.streams.tests.SmokeTestDriver.generate;
+import static org.apache.kafka.streams.tests.SmokeTestDriver.generatePerpetually;
+
+public class StreamsSmokeTest {
```
I'm assuming 22..25 client / drive code are all copy-pastes here so I skipped reviewing them. LMK if they aren't.

vvcephei replied:
That's correct.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/TimeOrderedKeyValueBufferChangelogDeserializationHelper.java:
```diff
@@ -0,0 +1,111 @@
+        return deserializationResult;
+    }
+
+    private static DeserializationResult deserializeV2(final ConsumerRecord<byte[], byte[]> record,
```
Some docs, either here or directly inside `InMemoryTimeOrderedKeyValueBuffer.java` explaining the format difference would help a lot. You can see some examples like `object GroupMetadataManager`

vvcephei replied:
sure thing!

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -258,34 +263,43 @@ private void logValue(final Bytes key, final BufferKey bufferKey, final BufferVa
+            changelogTopic,
+            key,
+            null,
+            null,
```
I'm just thinking, maybe we should encode headers to tombstones too in case in the future we changed the semantics of tombstones?

vvcephei replied:
I remember considering this when I added the first version header. The reason I didn't is that, since the initial version didn't have any headers, even if we change the tombstone format in the future, we'll always have to interpret a "no header, null value" record as being a "legacy format" tombstone, just like we have to interpret a "no header, non-null value" as being a "legacy format" data record.

You can think of "no header" as indicating "version 0". Since we haven't changed the format of tombstones _yet_, there's no value in adding a "version 1" flag. We should just wait until we do need to make such a change (if ever).

guozhangwang replied:
SG

guozhangwang reviewed (commented):
Triggered https://jenkins.confluent.io/view/all/job/system-test-kafka-branch-builder/3991/

guozhangwang commented:
test this

vvcephei commented:
Hmm. Still saw the OOME in https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/3134/

vvcephei commented:
Retest this please

vvcephei commented:
Ah, that heap space thing was legit. Fix coming...

Commit:
fix attempt to allocate arbitrary sized array

vvcephei commented:
Hey @guozhangwang , you might want to take a look at that last fix. The duck-typing code was producing an OOME some times, when it would just interpret a random integer out of the buffer as a "size" (integer) and blindly allocate an array of that size.

I added a Util (with tests) that has a guard to prevent this.

vvcephei commented:
I will follow up shortly to extract the system tests to a separate PR, since we're having trouble running the tests at all right now, and we wouldn't know if they are even more broken.

Commit:
factor out system test changes

Commit:
style

vvcephei commented:
Ok, @guozhangwang , This is my "final" iteration. I pulled the system tests out, and I'll follow up with another PR later. This PR should be sufficient for the basic purpose, thanks to the new "binary" compatibility unit tests.

guozhangwang reviewed (approved):
LGTM. We can merge after green build.

Let's trigger system test builds on the follow-up PR

vvcephei commented:
Failures were unrelated:

```
kafka.api.PlaintextConsumerTest > testLowMaxFetchSizeForRequestAndPartition FAILED
    org.scalatest.exceptions.TestFailedException: Timed out before consuming expected 2700 records. The number consumed was 1077.
        at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:530)
        at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:529)
        at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1389)
        at org.scalatest.Assertions.fail(Assertions.scala:1091)
        at org.scalatest.Assertions.fail$(Assertions.scala:1087)
        at org.scalatest.Assertions$.fail(Assertions.scala:1389)
        at kafka.api.AbstractConsumerTest.consumeRecords(AbstractConsumerTest.scala:158)
        at kafka.api.PlaintextConsumerTest.testLowMaxFetchSizeForRequestAndPartition(PlaintextConsumerTest.scala:804)
```

```
org.apache.kafka.connect.mirror.MirrorConnectorsIntegrationTest > testReplication FAILED
    java.lang.RuntimeException: Could not find enough records. found 0, expected 100
        at org.apache.kafka.connect.util.clusters.EmbeddedKafkaCluster.consume(EmbeddedKafkaCluster.java:435)
        at org.apache.kafka.connect.mirror.MirrorConnectorsIntegrationTest.testReplication(MirrorConnectorsIntegrationTest.java:217)
```

```
kafka.api.PlaintextConsumerTest.testLowMaxFetchSizeForRequestAndPartition
```

```
org.apache.kafka.streams.integration.HighAvailabilityTaskAssignorIntegrationTest.shouldScaleOutWithWarmupTasksAndPersistentStores
```

```
org.apache.kafka.connect.integration.BlockingConnectorTest.testBlockInConnectorStop
```

vvcephei commented:
backported to 2.6, 2.5, and 2.4. I ran the streams and client tests each time, as well as systemTestLibs.
